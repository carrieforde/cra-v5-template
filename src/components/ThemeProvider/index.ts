export * from "./ThemeProvider";
export * from "./useDarkMode";
