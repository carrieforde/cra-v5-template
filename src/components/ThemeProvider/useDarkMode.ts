import { useEffect, useState } from "react";

export const useDarkMode = () => {
  const darkModeMQ = window?.matchMedia("(prefers-color-scheme: dark");
  const [isDark, setIsDark] = useState<boolean>(darkModeMQ.matches || false);

  useEffect(() => {
    function handleColorSchemeChange(event: MediaQueryListEvent) {
      if (event.matches) {
        setIsDark(true);
      } else {
        setIsDark(false);
      }
    }

    darkModeMQ.addEventListener("change", handleColorSchemeChange);

    return () => {
      darkModeMQ.removeEventListener("change", handleColorSchemeChange);
    };
  }, [darkModeMQ]);

  return isDark;
};
