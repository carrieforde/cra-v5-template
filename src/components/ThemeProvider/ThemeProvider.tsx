import { ThemeProvider as EmotionThemeProvider } from "@emotion/react";
import { GlobalStyles } from "components/GlobalStyles";
import { FC, ReactNode } from "react";
import { buildCustomTheme, CustomThemeArgs } from "theme";
import { useDarkMode } from "./useDarkMode";

export type ThemeProviderProps = {
  children: ReactNode;
} & Omit<CustomThemeArgs, "useDarkMode">;

export const ThemeProvider: FC<ThemeProviderProps> = ({
  children,
  htmlFontSize,
  spacingSize,
  darkPalette,
}) => {
  const isDark = useDarkMode();
  const theme = buildCustomTheme({
    useDarkMode: isDark,
    htmlFontSize,
    spacingSize,
    darkPalette,
  });

  return (
    <EmotionThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </EmotionThemeProvider>
  );
};
