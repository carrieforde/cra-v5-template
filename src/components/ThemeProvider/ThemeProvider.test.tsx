import { render } from "@testing-library/react";
import { ThemeProvider } from "./ThemeProvider";

jest.mock("./useDarkMode", () => ({
  useDarkMode: jest.fn(() => false),
}));

describe("ThemeProvider", () => {
  it("should render the theme provider", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <p>Theme</p>
      </ThemeProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
