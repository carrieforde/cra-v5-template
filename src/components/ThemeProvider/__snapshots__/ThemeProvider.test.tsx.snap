// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider should render the theme provider 1`] = `
<DocumentFragment>
  <p>
    Theme
  </p>
</DocumentFragment>
`;
